### PR TITLE
reject non canonical encoding length.

### LINKED ISSUE #28: function check_length in x509v3.rs doesn't reject non canonical encoding length
For example, 
when the length encoding slice is : 0x81 0x12,
expected behavior: spdm-rs reject it and treat it as invalid length encoding,
current behavior: spdm-rs accept it and treat is as valid length encoding.

For reference: https://github.com/ccc-spdm-tools/spdm-rs/blob/3c678907cd3d121b8b2cc5d5048958b1201a8099/spdmlib/src/crypto/spdm_ring/cert_operation_impl.rs#L86-L135

### PR DESCRIPTION
**Pull request by longlongyang**
fix #28.

### CODE REVIEW

File: spdmlib/src/crypto/x509v3.rs [hunk 1]
```diff
@@ -263,19 +262,43 @@ fn check_length(data: &[u8]) -> SpdmResult<(usize, usize)> {
...
-        let length_count = data[0] - ASN1_LENGTH_MULTI_OCTET_MASK;
-        if len < (length_count as usize + 1) || length_count == 0 || length_count > 8 {
-            Err(SPDM_STATUS_VERIF_FAIL)
-        } else {
-            let mut length = [0u8; 8];
-            for (i, b) in data[1..length_count as usize + 1].iter().rev().enumerate() {
-                length[i] = *b;
+        let length_byte0 = data[0];
+        let (length, byte_comsumed) = match length_byte0 {
+            n if (n & 0x80) == 0 => (n as usize, 1),
```
Comment (jyao1) [inline spdmlib/src/crypto/x509v3.rs:268]: please don't use hardcoded 0x80 or 0x81.

File: spdmlib/src/crypto/x509v3.rs [hunk 2]
```diff
@@ -25,7 +25,6 @@ const ASN1_TAG_SEQUENCE: u8 =
     ASN1_TAG_CLASS_UNIVERSAL_MASK | ASN1_FORM_CONSTRUCTED_MASK | ASN1_TAG_NUMBER_SEQUENCE;
 const ASN1_TAG_EXPLICIT_EXTENSION: u8 = 0xA3;
 const ASN1_TAG_EXTN_VALUE: u8 = 0x04;
-const ASN1_LENGTH_MULTI_OCTET_MASK: u8 = 0x80;
```
Comment (jyao1) [inline spdmlib/src/crypto/x509v3.rs:28]: why remove it?

Comment (longlongyang) [reply spdmlib/src/crypto/x509v3.rs]: > why remove it?

updated.

File: spdmlib/src/crypto/x509v3.rs [hunk 3]
```diff
@@ -263,19 +263,43 @@ fn check_length(data: &[u8]) -> SpdmResult<(usize, usize)> {
...
-        if len < (length_count as usize + 1) || length_count == 0 || length_count > 8 {
-            Err(SPDM_STATUS_VERIF_FAIL)
-        } else {
-            let mut length = [0u8; 8];
-            for (i, b) in data[1..length_count as usize + 1].iter().rev().enumerate() {
-                length[i] = *b;
+        let length_byte0 = data[0];
+        let (length, byte_comsumed) = match length_byte0 {
+            n if (n & ASN1_LENGTH_MULTI_OCTET_MASK) == 0 => (n as usize, 1),
+            one_octet if one_octet == ASN1_LENGTH_MULTI_OCTET_MASK + 1 => {
```
Comment (jyao1) [inline spdmlib/src/crypto/x509v3.rs:270]: please define a new static variable.

File: spdmlib/src/crypto/x509v3.rs [hunk 4]
```diff
@@ -263,19 +263,43 @@ fn check_length(data: &[u8]) -> SpdmResult<(usize, usize)> {
...
+                    if second_byte < 0x80 {
+                        return Err(SPDM_STATUS_VERIF_FAIL); // Not the canonical encoding.
+                    } else {
+                        (second_byte, 2)
+                    }
+                }
             }
-            Ok((usize::from_le_bytes(length), length_count as usize + 1))
-        }
+            two_octet if two_octet == ASN1_LENGTH_MULTI_OCTET_MASK + 2 => {
```
Comment (jyao1) [inline spdmlib/src/crypto/x509v3.rs:282]: please define a new static variable.

File: spdmlib/src/crypto/x509v3.rs (hunk 3 above)
Comment (longlongyang) [reply spdmlib/src/crypto/x509v3.rs]: > please define a new static variable.

updated.

File: spdmlib/src/crypto/x509v3.rs (hunk 4 above)
Comment (longlongyang) [reply spdmlib/src/crypto/x509v3.rs]: > please define a new static variable.

updated.

**Pull request merged**